File: packages/jsii-pacmak/test/expected.jsii-calc/java/src/main/java/software/amazon/jsii/tests/calculator/StructB.java
package software.amazon.jsii.tests.calculator;

/**
 * This intentionally overlaps with StructA (where only requiredString is provided) to test htat the kernel properly disambiguates those.
 * 
 * EXPERIMENTAL
 */
@javax.annotation.Generated(value = "jsii-pacmak")
@software.amazon.jsii.Jsii(module = software.amazon.jsii.tests.calculator.$Module.class, fqn = "jsii-calc.StructB")
@software.amazon.jsii.Jsii.Proxy(StructB.Jsii$Proxy.class)
@software.amazon.jsii.Stability(software.amazon.jsii.Stability.Level.Experimental)
public interface StructB extends software.amazon.jsii.JsiiSerializable {

    /**
     * EXPERIMENTAL
     */
    @software.amazon.jsii.Stability(software.amazon.jsii.Stability.Level.Experimental)
    java.lang.String getRequiredString();

    /**
     * EXPERIMENTAL
     */
    @software.amazon.jsii.Stability(software.amazon.jsii.Stability.Level.Experimental)
    default java.lang.Boolean getOptionalBoolean() {
        return null;
    }

    /**
     * EXPERIMENTAL
     */
    @software.amazon.jsii.Stability(software.amazon.jsii.Stability.Level.Experimental)
    default software.amazon.jsii.tests.calculator.StructA getOptionalStructA() {
        return null;
    }

    /**
     * @return a {@link Builder} of {@link StructB}
     */
    @software.amazon.jsii.Stability(software.amazon.jsii.Stability.Level.Experimental)
    static Builder builder() {
        return new Builder();
    }
    /**
     * A builder for {@link StructB}
     */
    @software.amazon.jsii.Stability(software.amazon.jsii.Stability.Level.Experimental)
    public static final class Builder {
        private java.lang.String requiredString;
        private java.lang.Boolean optionalBoolean;
        private software.amazon.jsii.tests.calculator.StructA optionalStructA;

        /**
         * Sets the value of RequiredString
         * @param requiredString the value to be set. This parameter is required.
         * @return {@code this}
         */
        @software.amazon.jsii.Stability(software.amazon.jsii.Stability.Level.Experimental)
        public Builder requiredString(java.lang.String requiredString) {
            this.requiredString = requiredString;
            return this;
        }

        /**
         * Sets the value of OptionalBoolean
         * @param optionalBoolean the value to be set.
         * @return {@code this}
         */
        @software.amazon.jsii.Stability(software.amazon.jsii.Stability.Level.Experimental)
        public Builder optionalBoolean(java.lang.Boolean optionalBoolean) {
            this.optionalBoolean = optionalBoolean;
            return this;
        }

        /**
         * Sets the value of OptionalStructA
         * @param optionalStructA the value to be set.
         * @return {@code this}
         */
        @software.amazon.jsii.Stability(software.amazon.jsii.Stability.Level.Experimental)
        public Builder optionalStructA(software.amazon.jsii.tests.calculator.StructA optionalStructA) {
            this.optionalStructA = optionalStructA;
            return this;
        }

        /**
         * Builds the configured instance.
         * @return a new instance of {@link StructB}
         * @throws NullPointerException if any required attribute was not provided
         */
        @software.amazon.jsii.Stability(software.amazon.jsii.Stability.Level.Experimental)
        public StructB build() {
            return new Jsii$Proxy(requiredString, optionalBoolean, optionalStructA);
        }
    }

    /**
     * An implementation for {@link StructB}
     */
    @software.amazon.jsii.Stability(software.amazon.jsii.Stability.Level.Experimental)
    final class Jsii$Proxy extends software.amazon.jsii.JsiiObject implements StructB {
        private final java.lang.String requiredString;
        private final java.lang.Boolean optionalBoolean;
        private final software.amazon.jsii.tests.calculator.StructA optionalStructA;

        /**
         * Constructor that initializes the object based on values retrieved from the JsiiObject.
         * @param objRef Reference to the JSII managed object.
         */
        protected Jsii$Proxy(final software.amazon.jsii.JsiiObjectRef objRef) {
            super(objRef);
            this.requiredString = this.jsiiGet("requiredString", java.lang.String.class);
            this.optionalBoolean = this.jsiiGet("optionalBoolean", java.lang.Boolean.class);
            this.optionalStructA = this.jsiiGet("optionalStructA", software.amazon.jsii.tests.calculator.StructA.class);
        }

        /**
         * Constructor that initializes the object based on literal property values passed by the {@link Builder}.
         */
        private Jsii$Proxy(final java.lang.String requiredString, final java.lang.Boolean optionalBoolean, final software.amazon.jsii.tests.calculator.StructA optionalStructA) {
            super(software.amazon.jsii.JsiiObject.InitializationMode.JSII);
            this.requiredString = java.util.Objects.requireNonNull(requiredString, "requiredString is required");
            this.optionalBoolean = optionalBoolean;
            this.optionalStructA = optionalStructA;
        }

        @Override
        public java.lang.String getRequiredString() {
            return this.requiredString;
        }

        @Override
        public java.lang.Boolean getOptionalBoolean() {
            return this.optionalBoolean;
        }

        @Override
        public software.amazon.jsii.tests.calculator.StructA getOptionalStructA() {
            return this.optionalStructA;
        }

        @Override
        public com.fasterxml.jackson.databind.JsonNode $jsii$toJson() {
            final com.fasterxml.jackson.databind.ObjectMapper om = software.amazon.jsii.JsiiObjectMapper.INSTANCE;
            final com.fasterxml.jackson.databind.node.ObjectNode data = com.fasterxml.jackson.databind.node.JsonNodeFactory.instance.objectNode();

            data.set("requiredString", om.valueToTree(this.getRequiredString()));
            if (this.getOptionalBoolean() != null) {
                data.set("optionalBoolean", om.valueToTree(this.getOptionalBoolean()));
            }
            if (this.getOptionalStructA() != null) {
                data.set("optionalStructA", om.valueToTree(this.getOptionalStructA()));
            }

            final com.fasterxml.jackson.databind.node.ObjectNode struct = com.fasterxml.jackson.databind.node.JsonNodeFactory.instance.objectNode();
            struct.set("fqn", om.valueToTree("jsii-calc.StructB"));
            struct.set("data", data);

            final com.fasterxml.jackson.databind.node.ObjectNode obj = com.fasterxml.jackson.databind.node.JsonNodeFactory.instance.objectNode();
            obj.set("$jsii.struct", struct);

            return obj;
        }

        @Override
        public boolean equals(Object o) {
            if (this == o) return true;
            if (o == null || getClass() != o.getClass()) return false;

            StructB.Jsii$Proxy that = (StructB.Jsii$Proxy) o;

            if (!requiredString.equals(that.requiredString)) return false;
            if (this.optionalBoolean != null ? !this.optionalBoolean.equals(that.optionalBoolean) : that.optionalBoolean != null) return false;
            return this.optionalStructA != null ? this.optionalStructA.equals(that.optionalStructA) : that.optionalStructA == null;
        }

        @Override
        public int hashCode() {
            int result = this.requiredString.hashCode();
            result = 31 * result + (this.optionalBoolean != null ? this.optionalBoolean.hashCode() : 0);
            result = 31 * result + (this.optionalStructA != null ? this.optionalStructA.hashCode() : 0);
            return result;
        }
    }
}

File: packages/jsii-pacmak/test/expected.jsii-calc/kotlin/src/main/kotlin/software/amazon/jsii/tests/kotlin/calculator/StructB.kt
package software.amazon.jsii.tests.kotlin.calculator

/**
 * This intentionally overlaps with StructA (where only requiredString is provided) to test htat the kernel properly disambiguates those.
 */
@javax.annotation.Generated("jsii-pacmak")
@software.amazon.jsii.Stability(software.amazon.jsii.Stability.Level.Experimental)
interface StructB : software.amazon.jsii.JsiiSerializable {
    @get:software.amazon.jsii.Stability(software.amazon.jsii.Stability.Level.Experimental)
    val requiredString: kotlin.String

    @get:software.amazon.jsii.Stability(software.amazon.jsii.Stability.Level.Experimental)
    val optionalBoolean: kotlin.Boolean?

    @get:software.amazon.jsii.Stability(software.amazon.jsii.Stability.Level.Experimental)
    val optionalStructA: software.amazon.jsii.tests.kotlin.calculator.StructA?

    class Builder {
        @get:software.amazon.jsii.Stability(software.amazon.jsii.Stability.Level.Experimental)
        var requiredString: kotlin.String? = null

        @get:software.amazon.jsii.Stability(software.amazon.jsii.Stability.Level.Experimental)
        var optionalBoolean: kotlin.Boolean? = null

        @get:software.amazon.jsii.Stability(software.amazon.jsii.Stability.Level.Experimental)
        var optionalStructA: software.amazon.jsii.tests.kotlin.calculator.StructA? = null

        fun build(): software.amazon.jsii.tests.kotlin.calculator.StructB {
            val requiredString = this.requiredString ?: kotlin.error("Value for property 'requiredString' must be specified")
            val optionalBoolean = this.optionalBoolean
            val optionalStructA = this.optionalStructA
            return `Jsii$Proxy`(requiredString, optionalBoolean, optionalStructA)
        }
    }

    class `Jsii$Proxy` : software.amazon.jsii.JsiiObject, software.amazon.jsii.tests.kotlin.calculator.StructB {
        @get:software.amazon.jsii.Stability(software.amazon.jsii.Stability.Level.Experimental)
        override val requiredString: kotlin.String

        @get:software.amazon.jsii.Stability(software.amazon.jsii.Stability.Level.Experimental)
        override val optionalBoolean: kotlin.Boolean?

        @get:software.amazon.jsii.Stability(software.amazon.jsii.Stability.Level.Experimental)
        override val optionalStructA: software.amazon.jsii.tests.kotlin.calculator.StructA?

        protected constructor(objRef: software.amazon.jsii.JsiiObjectRef) : super(objRef) {
            this.requiredString = this.jsiiCall("requiredString", kotlin.String::class.java) ?: error("'requiredString' should be present")
            this.optionalBoolean = this.jsiiCall("optionalBoolean", kotlin.Boolean::class.java)
            this.optionalStructA = this.jsiiCall("optionalStructA", software.amazon.jsii.tests.kotlin.calculator.StructA::class.java)
        }

        constructor(requiredString: kotlin.String, optionalBoolean: kotlin.Boolean?, optionalStructA: software.amazon.jsii.tests.kotlin.calculator.StructA?) : super(software.amazon.jsii.JsiiObject.InitializationMode.JSII) {
            this.requiredString = requiredString
            this.optionalBoolean = optionalBoolean
            this.optionalStructA = optionalStructA
        }

        override fun equals(other: Any?): Boolean {
            if (this === other) return true
            if (javaClass != other?.javaClass) return false
            other as software.amazon.jsii.tests.kotlin.calculator.StructB
            if (requiredString != other.requiredString) return false
            if (optionalBoolean != other.optionalBoolean) return false
            if (optionalStructA != other.optionalStructA) return false
            return true
        }

        override fun hashCode(): Int {
            var result = this.requiredString.hashCode()
            result = 31 * result + (this.optionalBoolean?.hashCode() ?: 0)
            result = 31 * result + (this.optionalStructA?.hashCode() ?: 0)
            return result
        }

        override fun `$jsii$toJson`(): com.fasterxml.jackson.databind.JsonNode {
            val om = software.amazon.jsii.JsiiObjectMapper.INSTANCE
            val obj = com.fasterxml.jackson.databind.node.JsonNodeFactory.instance.objectNode()
            obj.set<com.fasterxml.jackson.databind.JsonNode>("requiredString", om.valueToTree(this.requiredString))
            obj.set<com.fasterxml.jackson.databind.JsonNode>("optionalBoolean", om.valueToTree(this.optionalBoolean))
            obj.set<com.fasterxml.jackson.databind.JsonNode>("optionalStructA", om.valueToTree(this.optionalStructA))
            return obj
        }
    }
}
